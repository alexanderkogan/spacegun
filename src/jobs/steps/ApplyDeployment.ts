import { call } from "../../dispatcher"
import { IO } from "../../IO"

import * as clusterModule from "../../cluster/ClusterModule"

import { DeploymentPlan } from "../model/DeploymentPlan"
import { Deployment } from "../../cluster/model/Deployment"

export class ApplyDeployment {

    private readonly io: IO = new IO()

    public async apply(plan: DeploymentPlan): Promise<Deployment> {
        const deployment = await call(clusterModule.updateDeployment)(plan)
        this.io.out(`sucessfully updated ${deployment.name} with image ${deployment.image!}`)
        return deployment
    }
}
