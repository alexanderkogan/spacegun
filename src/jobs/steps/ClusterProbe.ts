import axios from "axios"

export class ClusterProbe {

    public async apply(input: object, hook: string, timeout?: number): Promise<object> {
        const result = await axios.get(hook, { timeout })
        if (result.status === 200) {
            return Promise.resolve(input)
        } else {
            return Promise.reject(new Error(
                `Cluster probe ${hook} returned status code ${result.status}`
            ))
        }
    }
}
