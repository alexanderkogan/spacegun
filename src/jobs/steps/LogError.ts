import { call } from "../../dispatcher"
import { IO } from "../../IO"

import * as eventModule from "../../events/EventModule"
import { PipelineDescription } from "../model/PipelineDescription";

export class LogError {

    public constructor(
        public readonly io: IO = new IO()
    ) { }

    public async apply(pipeline: PipelineDescription, error: Error): Promise<{}> {
        this.io.error(error)
        call(eventModule.log)({
            message: `Failed to apply pipeline ${pipeline.name}`,
            timestamp: Date.now(),
            topics: ["slack"],
            description: `An error occured during pipeline execution: ${error.message}`,
            fields: [{
                title: "Stack",
                value: error.stack || ""
            }]
        })
        return {}
    }
}
