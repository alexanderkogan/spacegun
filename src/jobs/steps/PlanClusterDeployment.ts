import { IO } from '../../IO'

import { call } from '../../dispatcher'

import * as clusterModule from '../../cluster/ClusterModule'
import { ServerGroup } from '../../cluster/model/ServerGroup'
import { Deployment } from '../../cluster/model/Deployment'
import { Filter, matchesServerGroup, matchesDeployment } from '../model/Filter'

import { DeploymentPlan } from '../model/DeploymentPlan'
import { JobPlan } from '../model/JobPlan'

export class PlanClusterDeployment {
    public constructor(readonly name: string, readonly cluster: string, readonly filter?: Partial<Filter>, readonly io: IO = new IO()) {}

    public async plan(group: ServerGroup, name: string, targetDeployments: Deployment[]): Promise<JobPlan> {
        if (!matchesServerGroup(this.filter, group)) {
            return { name, deployments: [] }
        }

        const sourceDeployments = await call(clusterModule.deployments)({
            cluster: this.cluster!,
            namespace: group.namespace,
        })

        const deployments: DeploymentPlan[] = []
        for (const targetDeployment of targetDeployments) {
            if (!matchesDeployment(this.filter, targetDeployment)) {
                continue
            }
            this.io.out(`planning cluster deployment ${targetDeployment.name} in ${this.name}`)
            const sourceDeployment = sourceDeployments.find(d => d.name === targetDeployment.name)
            if (sourceDeployment === undefined) {
                this.io.error(
                    `${targetDeployment.name} in cluster ${group.cluster} has no appropriate deployment in cluster ${this.cluster}`
                )
                continue
            }
            if (sourceDeployment.image === undefined) {
                this.io.error(`${targetDeployment.name} in cluster ${group.cluster} has no image`)
                continue
            }
            if (targetDeployment.image === undefined || targetDeployment.image.url !== sourceDeployment.image.url) {
                deployments.push({
                    group,
                    deployment: targetDeployment,
                    image: sourceDeployment.image,
                })
            }
        }
        return { name, deployments }
    }
}
