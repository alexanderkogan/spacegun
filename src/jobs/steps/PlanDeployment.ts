import { Pipe } from "lazy-space"
import { Job } from "@/jobs/model/Job"
import { ServerGroup } from "@/cluster/model/ServerGroup"
import { IO } from "@/IO"

import { call } from "@/dispatcher"

import * as clusterModule from "@/cluster/ClusterModule"
import * as imageModule from "@/images/ImageModule"

import { Deployment } from "@/cluster/model/Deployment"
import { DeploymentPlan } from "@/jobs/model/DeploymentPlan"
import { Eval, PromiseEval } from "lazy-space/lib/eval"

interface PlanInput {
    job: Job
    namespace?: string
}

export class PlanDeployment extends Pipe<PlanInput, DeploymentPlan[]> {

    private readonly io: IO = new IO()

    public pass(input: PlanInput): Eval<DeploymentPlan[]> {
        const job = input.job
        const namespace = input.namespace

        return new PromiseEval(call(clusterModule.deployments)({ cluster: job.cluster, namespace }))
            .flatMap(targetDeployments => {
                const group: ServerGroup = { cluster: job.cluster, namespace: namespace }

                switch (input.job.from.type) {
                    case "cluster": return new PromiseEval(this.planClusterDeployment(job, targetDeployments, group))
                    case "image": return new PromiseEval(this.planImageDeployment(job, targetDeployments, group))
                }
            })
    }

    async planClusterDeployment(job: Job, targetDeployments: Deployment[], group: ServerGroup): Promise<DeploymentPlan[]> {
        const deployments: DeploymentPlan[] = []
        const sourceDeployments = await call(clusterModule.deployments)({
            cluster: job.from.expression!,
            namespace: group.namespace
        })
        for (const targetDeployment of targetDeployments) {
            this.io.out(`planning cluster deployment ${targetDeployment.name} in job ${job.name}`)
            const sourceDeployment = sourceDeployments.find(d => d.name === targetDeployment.name)
            if (sourceDeployment === undefined) {
                console.error(`${targetDeployment.name} in cluster ${group.cluster} has no appropriate deployment in cluster ${job.from.expression}`)
                continue
            }
            if (sourceDeployment.image === undefined) {
                console.error(`${targetDeployment.name} in cluster ${group.cluster} has no image`)
                continue
            }
            if (targetDeployment.image === undefined || targetDeployment.image.url !== sourceDeployment.image.url) {
                deployments.push({
                    group,
                    deployment: targetDeployment,
                    image: sourceDeployment.image
                })
            }
        }
        return deployments
    }

    async planImageDeployment(job: Job, targetDeployments: Deployment[], group: ServerGroup): Promise<DeploymentPlan[]> {
        const deployments: DeploymentPlan[] = []

        for (const targetDeployment of targetDeployments) {
            this.io.out(`planning image deployment ${targetDeployment.name} in job ${job.name}`)
            if (targetDeployment.image === undefined) {
                console.error(`${targetDeployment.name} in cluster ${group.cluster} has no image, so spacegun cannot determine the right image source`)
                continue
            }
            const image = await call(imageModule.image)({
                tag: job.from.expression,
                name: targetDeployment.image.name
            })
            if (targetDeployment.image.url !== image.url) {
                deployments.push({
                    group,
                    image,
                    deployment: targetDeployment,
                })
            }
        }
        return deployments
    }
}
