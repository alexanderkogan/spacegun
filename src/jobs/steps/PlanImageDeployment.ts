import { IO } from "../../IO"

import { call } from "../../dispatcher"

import * as imageModule from "../../images/ImageModule"

import { DeploymentPlan } from "../model/DeploymentPlan"
import { Deployment } from "../../cluster/model/Deployment"
import { ServerGroup } from "../../cluster/model/ServerGroup"
import { Filter, matchesDeployment, matchesServerGroup } from "../model/Filter"
import { JobPlan } from "../model/JobPlan";

export type FetchedDeployment = {}

export class PlanImageDeployment {

    public constructor(
        readonly name: string,
        readonly tag: string | undefined,
        readonly semanticTagExtractor: RegExp | string | undefined,
        readonly filter?: Partial<Filter>,
        readonly io: IO = new IO()
    ) { }

    public async plan(group: ServerGroup, name: string, targetDeployments: Deployment[]): Promise<JobPlan> {
        if (!matchesServerGroup(this.filter, group)) {
            return { name, deployments: [] }
        }

        const deployments: DeploymentPlan[] = []
        for (const targetDeployment of targetDeployments) {
            if (!matchesDeployment(this.filter, targetDeployment)) {
                continue
            }
            this.io.out(`planning image deployment ${targetDeployment.name} in ${this.name}`)
            if (targetDeployment.image === undefined) {
                this.io.error(`${targetDeployment.name} in cluster ${group.cluster} has no image, so spacegun cannot determine the right image source`)
                continue
            }

            let tag = this.tag
            if (tag === undefined) {
                const tags = await call(imageModule.tags)(targetDeployment.image)
                tag = this.getNewestTag(tags)
            }

            if (tag !== undefined) {
                const image = await call(imageModule.image)({
                    tag, name: targetDeployment.image.name
                })

                if (targetDeployment.image.url !== image.url) {
                    deployments.push({
                        group,
                        image,
                        deployment: targetDeployment,
                    })
                }
            } else {
                throw new Error(`Could not find a tag for deployment ${targetDeployment.name} in cluster ${group.cluster}, namespace ${group.namespace}`)
            }
        }
        return { name, deployments }
    }

    private getNewestTag(tags: string[]): string | undefined {
        let sortableTags: string[]
        if (this.semanticTagExtractor !== undefined) {
            sortableTags = tags.map(tag => tag.match(this.semanticTagExtractor!))
                .filter(match => match !== null && match.length > 0)
                .map(match => match![0])
        } else {
            sortableTags = tags
        }

        sortableTags.sort((a, b) => b.localeCompare(a))
        if (sortableTags.length === 0) {
            return undefined
        }
        if (sortableTags.length > 1 && sortableTags[0] === sortableTags[1]) {
            return undefined
        }
        return sortableTags.length > 0 ? sortableTags[0] : undefined
    }
}
