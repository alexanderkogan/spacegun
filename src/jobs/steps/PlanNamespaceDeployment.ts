import { call } from '../../dispatcher'
import { IO } from '../../IO'

import { ServerGroup } from '../../cluster/model/ServerGroup'
import { Deployment } from '../../cluster/model/Deployment'
import * as clusterModule from '../../cluster/ClusterModule'

import { Filter, matchesServerGroup, matchesResource } from '../model/Filter'
import { JobPlan } from '../model/JobPlan'
import { DeploymentPlan, DeployableResource } from '../model/DeploymentPlan'
import { Batch } from 'src/cluster/model/Batch'

export class PlanNamespaceDeployment {
    public constructor(
        readonly name: string,
        readonly sourceCluster: string,
        readonly sourceNamespace: string,
        readonly targetNamespace: string,
        readonly filter?: Partial<Filter>,
        readonly io: IO = new IO()
    ) {}

    public async plan(targetGroup: ServerGroup, name: string, targetDeployments: Deployment[], targetBatches: Batch[]): Promise<JobPlan> {
        if (!matchesServerGroup({ namespaces: [this.targetNamespace] }, targetGroup)) {
            return { name, deployments: [], batches: [] }
        }

        const deployments = await this.planDeployments(targetGroup, targetDeployments)
        const batches = await this.planBatches(targetGroup, targetBatches)

        return { name, deployments, batches }
    }

    async planDeployments(group: ServerGroup, targetDeployments: Deployment[]): Promise<DeploymentPlan<Deployment>[]> {
        const sourceDeployments = await call(clusterModule.deployments)({
            cluster: this.sourceCluster,
            namespace: this.sourceNamespace,
        })
        return this.planUpdates(group, sourceDeployments, targetDeployments)
    }

    async planBatches(group: ServerGroup, targetBatches: Batch[]): Promise<DeploymentPlan<Batch>[]> {
        const sourceBatches = await call(clusterModule.batches)({
            cluster: this.sourceCluster,
            namespace: this.sourceNamespace,
        })
        return this.planUpdates(group, sourceBatches, targetBatches)
    }

    async planUpdates<T extends DeployableResource>(group: ServerGroup, sources: T[], targets: T[]): Promise<DeploymentPlan<T>[]> {
        const deployments: DeploymentPlan<T>[] = []
        for (const target of targets) {
            if (!matchesResource(this.filter, target)) {
                continue
            }
            this.io.out(`planning namespace deployment ${target.name} in ${this.name}`)
            const source = sources.find(d => d.name === target.name)
            if (source === undefined) {
                this.io.error(
                    `${target.name} in {cluster: ${group.cluster}, namespace: ${group.namespace}} ` +
                        `has no appropriate deployment in {cluster: ${this.sourceCluster}, namespace: ${this.sourceNamespace}}`
                )
                continue
            }
            if (source.image === undefined) {
                this.io.error(`${target.name} in {cluster: ${this.sourceCluster}, namespace: ${this.sourceNamespace}} ` + `has no image`)
                continue
            }
            if (target.image === undefined || target.image.url !== source.image.url) {
                deployments.push({
                    group,
                    deployable: target,
                    image: source.image,
                })
            }
        }
        return deployments
    }
}
