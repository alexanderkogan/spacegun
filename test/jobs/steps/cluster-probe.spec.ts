import { ClusterProbe } from "../../../src/jobs/steps/ClusterProbe"

import axios from "axios"
import { axiosResponse } from "../../test-utils/axios"

describe(ClusterProbe.name, () => {

    it("resolves to input cluster probe returns status 200", async () => {
        // given
        axios.get = axiosResponse(200)

        // when
        const result = await new ClusterProbe().apply({ input: "" }, "hookurl")

        // then
        expect(result).toEqual({ input: "" })
    })

    it("rejects if cluster probe return status not 200", async () => {
        axios.get = axiosResponse(201)
        await expect(new ClusterProbe().apply({ input: "" }, "hookurl")).rejects.toThrowErrorMatchingSnapshot()
    })

    it("rejects if cluster probe rejects", async () => {
        axios.get = jest.fn().mockReturnValue(Promise.reject("timeout"))
        await expect(new ClusterProbe().apply({ input: "" }, "hookurl")).rejects.toEqual("timeout")
    })

    it("calls axios with optional timeout", async () => {
        axios.get = axiosResponse(200)
        await new ClusterProbe().apply({ input: "" }, "hookurl", 1234)
        expect(axios.get).toHaveBeenCalledWith("hookurl", { timeout: 1234 })
    })
})
