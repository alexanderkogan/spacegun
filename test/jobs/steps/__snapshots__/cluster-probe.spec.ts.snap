// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ClusterProbe rejects if cluster probe return status not 200 1`] = `"Cluster probe hookurl returned status code 201"`;
